### PR TITLE
config/fcos: Add extensions param and generate treefile with packages

### PR DESCRIPTION
**Pull request by jmarrero**
This PR adds the sugar to take a list of extensions (packages) on the butane config and turns them into a [treefile](https://coreos.github.io/rpm-ostree/treefile/) which rpm-ostree will consume as part of the derivation/layering [enhancement](https://github.com/coreos/enhancements/pull/7). This specific PR is raised based on the discussions on: https://github.com/coreos/fedora-coreos-tracker/issues/1054

### CODE REVIEW

**Comment by jmarrero** [comment]
@bgilbert is the way I marshaling the YAML correct. It's ok to add that dependency of `yaml.v2@v2.2.2` ?

**Comment by jmarrero** [comment]
Still seeing: ```Missing paths in TranslationSet:
$.storage.files.2.path
$.storage.files.2.contents.source
$.storage.files.2.contents
$.storage.files.2.mode
$.storage.files.2``` after merges trying to figure it out.

**Comment by bgilbert** [comment]
> is the way I am marshaling the YAML correct? It's OK to add that dependency of yaml.v2@v2.2.2 or am I missing some obvious util?

We already have a dependency on `yaml.v3`; can you use that?

> Missing paths in TranslationSet:

What does `butane -D` say?

**Comment by jmarrero** [comment]
> > is the way I am marshaling the YAML correct? It's OK to add that dependency of yaml.v2@v2.2.2 or am I missing some obvious util?
> 
> We already have a dependency on `yaml.v3`; can you use that?
> 
Yeah yaml.v3 works, I did not think of searching the other versions in the imports we are already using. Thanks.

> > Missing paths in TranslationSet:
> 
> What does `butane -D` say?


That output comes from butane -D.
The full output looks like:
```
TranslationSet: yaml → json
$.ignition → $.ignition
$.version → $.ignition.version
$.storage → $.storage
$.storage.files → $.storage.files
$.storage.files.0 → $.storage.files.0
$.storage.files.0.contents → $.storage.files.0.contents
$.inline → $.storage.files.0.contents.source
$.storage.files.0.mode → $.storage.files.0.mode
$.storage.files.0.path → $.storage.files.0.path
$.storage.files.1 → $.storage.files.1
$.storage.files.1.contents → $.storage.files.1.contents
$.storage.files.1.contents.inline → $.storage.files.1.contents.compression
$.storage.files.1.contents.inline → $.storage.files.1.contents.source
$.storage.files.1.mode → $.storage.files.1.mode
$.storage.files.1.path → $.storage.files.1.path
$.systemd → $.systemd
$.systemd.units → $.systemd.units
$.systemd.units.0 → $.systemd.units.0
$.systemd.units.0.dropins → $.systemd.units.0.dropins
$.systemd.units.0.dropins.0 → $.systemd.units.0.dropins.0
$.systemd.units.0.dropins.0.contents → $.systemd.units.0.dropins.0.contents
$.systemd.units.0.dropins.0.name → $.systemd.units.0.dropins.0.name
$.systemd.units.0.name → $.systemd.units.0.name
$.systemd.units.1 → $.systemd.units.1
$.systemd.units.1.contents → $.systemd.units.1.contents
$.systemd.units.1.enabled → $.systemd.units.1.enabled
$.systemd.units.1.name → $.systemd.units.1.name

Missing paths in TranslationSet:
$.storage.files.2.path
$.storage.files.2.contents.source
$.storage.files.2.contents
$.storage.files.2.mode
$.storage.files.2
```

Which if I understand is because the translation set from the procressPackages() sets the file at index 0 but after appending it is of index 2. If that is correct do I need to modify that manually or is there other method I am missing that is better than merge to deal with this?

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:86]: You can't just glue some arrays together here; you need to run a proper config merge with `MergeTranslatedConfigs`.  That'll take care of mapping the translation set indices for you.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:307]: In struct translation functions, the path is relative to the root of the struct.  In desugaring functions, though, it's relative to the root of the config.  So this isn't correct; you need a full path here.

**Review by bgilbert — COMMENTED** [review]
You'll want `MergeTranslatedConfigs`; see below.

File: config/fcos/v1_5_exp/translate.go
Comment (jmarrero) [reply config/fcos/v1_5_exp/translate.go]: I am not entirely sure I get which config. My latest commit I assumes is the config the user provides not the types.Config right?. So `yamlPath := path.New("yaml", "extensions")` ?

Comment (bgilbert) [reply config/fcos/v1_5_exp/translate.go]: Right, the root of the input (YAML, Butane) config.

**Comment by jmarrero** [comment]
Working on a test now. Thanks for all the help @bgilbert

File: config/fcos/v1_5_exp/schema.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/schema.go:24]: `Extensions`

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:24]: This block is for Butane packages.  External dependencies should be in the next one.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:306]: Whoops, I had the return order wrong.  Looks like the report should come last.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:325]: Could be simplified to an anonymous struct:

```go
	treeFileContents, err := yaml.Marshal(&struct {
		Packages []string `yaml:"packages"`
	}{
		Packages: c.Extensions,
	})
```

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:336]: Cleanup: the path literal could be inlined here.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:353]: This fails to add paths for `storage` and `storage.files`.
```suggestion
	ts.AddFromCommonSource(yamlPath, path.New("json", "storage"), ret.Storage)
```

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:89]: Cleanup: let's keep all of these lines in one block to show they're related.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:85]: I'd be tempted to spell this

```go
var retConfig types.Config
retConfig, ts = ...
```

but I can see it going either way.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:340]: The second argument is always nil, so for clarity I think it's better to pass nil there.

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:350]: Cleanup: I'd probably move the `file` construction down here, and set the fixed fields as part of construction.

File: config/fcos/v1_5_exp/translate.go
Comment (jmarrero) [reply config/fcos/v1_5_exp/translate.go:340]: If we know that it's always going be nil and gzipped is gonna always be false, any reason to not just ignore `gzipped` and turn it into `_`?

Comment (bgilbert) [reply config/fcos/v1_5_exp/translate.go:340]: `gzipped` won't always be false.  `MakeDataURL` makes a decision about whether to compress.

Comment (bgilbert) [reply config/fcos/v1_5_exp/translate.go:340]: The actual rule is: skip compression if the third argument is `false` or if the second argument is `&"gzip"`.

**Comment by jmarrero** [comment]
Moving back to WIP. This depends on the outcome of: 
https://github.com/coreos/rpm-ostree/pull/3340

File: config/fcos/v1_5_exp/schema.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/schema.go:43]: Are we sure we won't have additional options related to extensions in the future?  This schema doesn't allow for any.

The usual way we do this is:
```yaml
extensions:
  - name: foo
  - name: bar
```

File: config/fcos/v1_5_exp/translate.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate.go:322]: Bikeshed:
```suggestion
	fullYamlContents := append([]byte("# Generated by Butane\n\n"), treeFileContents...)
```

**Review by bgilbert — COMMENTED** [review]
LGTM generally, pending tests.

File: config/fcos/v1_5_exp/schema.go
Comment (jmarrero) [reply config/fcos/v1_5_exp/schema.go:43]: We are basically adding something like the RHCOS extensions to butane: https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfiguration.md#rhcos-extensions .  But I am not sure how this will evolve. This is still being discussed but for the first pass AFAIK we just care about a list of packages.

File: config/fcos/v1_5_exp/translate_test.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate_test.go:1429]: This is not an empty config.  :slightly_smiling_face:

File: docs/config-fcos-v1_5-exp.md
Comment (bgilbert) [inline docs/config-fcos-v1_5-exp.md:206]: lowercase A.

Also, we should either a) add this to the OpenShift experimental spec docs, or (b) have the OpenShift spec validator explicitly reject the field.

File: config/fcos/v1_5_exp/schema.go
Comment (bgilbert) [reply config/fcos/v1_5_exp/schema.go:43]: Once this spec version has been stabilized, we lose the ability to add metadata in this way without making a new top-level field.  We could merge this as-is and revisit before stabilization, but if we do, we should make sure to leave clear signposts in the issue tracker so we don't forget.  Or we could go for the extensible approach now.  It might be worth raising with the team.

File: config/fcos/v1_5_exp/translate_test.go
Comment (jmarrero) [reply config/fcos/v1_5_exp/translate_test.go:1429]: Wooops. Left behind from copy/paste from the original test. 😀 addressing asap

File: docs/config-fcos-v1_5-exp.md
Comment (jmarrero) [reply docs/config-fcos-v1_5-exp.md:206]: Is there a best practice on which to select? Specially if this could be changing?

Comment (bgilbert) [reply docs/config-fcos-v1_5-exp.md:206]: It's safer to disable the functionality in `openshift` and possibly enable it later, rather than the other way around.

Now that I think about it, `openshift` configs already have a way to specify OS extensions (`openshift.extensions`).  The last time we added that sort of duplication (for kernel arguments) we had `openshift` disable the fields inherited from the `fcos` variant, and continued to support only the `openshift`-specific fields there.

File: config/fcos/v1_5_exp/schema.go
Comment (jmarrero) [reply config/fcos/v1_5_exp/schema.go:43]: @cgwalters @jlebon any preference on how to handle this?

File: docs/config-fcos-v1_5-exp.md
Comment (jmarrero) [reply docs/config-fcos-v1_5-exp.md]: This might tie into this thread: https://github.com/coreos/butane/pull/304#discussion_r790336682 we want this to be fcos + openshift thing. So it might require another chat with the MCO team. I am more leaning into creating a Issue that says to revisit this once it's time to stabilize the API. I think for now we need to allow both on openshift.

Comment (bgilbert) [reply docs/config-fcos-v1_5-exp.md]: Well, it also ties into https://github.com/coreos/butane/issues/300#issuecomment-1015041837.  Even if `openshift` inherits it now, that'll change when 4.11.0 is stabilized unless the MCO starts supporting Ignition spec 3.3.0.

So I'm okay leaving it enabled for now and filing a bug.  I've created stabilization milestones for the next `fcos` and `openshift` specs in the issue tracker, and we can attach to those.  Could you handle filing the issue?

File: config/fcos/v1_5_exp/translate_test.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate_test.go:1468]: e83e1c99d8e45b7f8e0add5dffaef9751d34370f introduced a style change for these loops.  Could you apply the update when you next respin?

File: config/fcos/v1_5_exp/translate_test.go
Comment (bgilbert) [inline config/fcos/v1_5_exp/translate_test.go:1471]: Copypasted typo: `mismatch`.  I'll handle fixing the original.

File: docs/config-fcos-v1_5-exp.md
Comment (bgilbert) [reply docs/config-fcos-v1_5-exp.md]: In that case, please do update the `openshift` spec docs as well.

File: config/fcos/v1_5_exp/translate_test.go
Comment (bgilbert) [reply config/fcos/v1_5_exp/translate_test.go:1471]: #313

File: config/fcos/v1_5_exp/schema.go
Comment (jlebon) [reply config/fcos/v1_5_exp/schema.go:43]: Agree with the concern, but not sure what to recommend since it's hard to predict what we'll need to generalize (container-based extensions? more treefile support? repo-pinned extensions?).

So... I'd say let's roll with this for now and revisit before stabilization. In the worst case, we should be able to change the spec in future versions, right?

File: docs/config-fcos-v1_5-exp.md
Comment (jmarrero) [reply docs/config-fcos-v1_5-exp.md]: Opened: https://github.com/coreos/butane/issues/314

File: config/fcos/v1_5_exp/translate_test.go
Comment (jmarrero) [reply config/fcos/v1_5_exp/translate_test.go:1471]: done

File: config/fcos/v1_5_exp/schema.go
Comment (bgilbert) [reply config/fcos/v1_5_exp/schema.go:43]: Discussed this more with @jlebon OOB.  If we wanted to add extensibility in future versions, and didn't want to have a major spec bump (i.e. one that requires config changes when bumping the version number), we'd need to deprecate the `extensions` field and add e.g. `extensions2`.  We've done that before but it's a bit awkward.

I think the maximum possible extensibility is:

```yaml
extensions:
  packages:
    - name: foo
    - name: bar
```

which lets us add system-wide extension options, per-package ones, and new extension types:

```yaml
extensions:
  global-option: z
  packages:
    - name: foo
      package-option: z
    - name: bar
  containers:
    - name: baz
```

Global options don't play well with config merging, since a child config that adds a package would be affected by global parameters it isn't aware of.  New extension types could be implemented as a package field:

```yaml
extensions:
  - name: foo
    package-option: z
  - name: bar
  - name: baz
    type: container
```

which is consistent with the design elsewhere in the schema.

I think that's a reasonable middle ground.  Let's not block the current PR on this, but as a followup I think we should switch from an array of strings to an array of structs.

**Review by bgilbert — APPROVED** [review]

File: config/fcos/v1_5_exp/schema.go
Comment (cgwalters) [reply config/fcos/v1_5_exp/schema.go:43]: This is very related to what happened to `packages` in 
https://github.com/coreos/rpm-ostree/blob/main/docs/treefile.md

We ended up adding `repo-packages` which allows to pin to packages from a specific repo.

--

The idea of `type: container` is a really interesting idea.  Definitely touches on https://www.freedesktop.org/software/systemd/man/systemd-sysext.html etc.

**Comment by cgwalters** [comment]
One concern @jmarrero mentioned in a chat about this is that we're implicity supporting whatever CLI arguments can be passed to `rpm-ostree install`.  Today that includes e.g. bodhi update URLs as well as direct e.g. `https://example.com/foo.rpm`.

I think that's probably OK.

This also relates though to a discussion happening in libdnf 
https://github.com/rpm-software-management/libdnf/issues/1375

**Comment by cgwalters** [comment]
Did the question come up of potentially having butane generate default configuration files or kernel arguments when a specific extension is requested?  `kdump` is a great example of this.  https://docs.fedoraproject.org/en-US/fedora-coreos/debugging-kernel-crashes/

There may be others where we'd want to drop some config file in `/etc`.

IOW, should butane ever try to recognize specific extensions, or should we just pass through to `rpm-ostree install`.

File: config/fcos/v1_5_exp/schema.go
Comment (cgwalters) [reply config/fcos/v1_5_exp/schema.go:43]: In rpm-ostree's treefiles today to be nice we allow multiple packages in single YAML list entry.  See for example https://github.com/coreos/fedora-coreos-config/blob/aa50439847f34282b41bf2eb60d478a45d2a27d7/manifests/system-configuration.yaml#L7

Might make sense to do that too to avoid the overhead in the simple cases?

Comment (bgilbert) [reply config/fcos/v1_5_exp/schema.go:43]: We generally don't do that in _Ignition_ configs because it wouldn't work with the config merging logic, which (with one unfortunate exception) doesn't know the semantics of what it's merging.  That argument doesn't directly apply here, since this is sugar, but it still doesn't feel very Butane-like.  We've favored explicitness over ergonomics when they're in conflict.

**Comment by bgilbert** [comment]
> One concern @jmarrero mentioned in a chat about this is that we're implicity supporting whatever CLI arguments can be passed to rpm-ostree install. Today that includes e.g. bodhi update URLs as well as direct e.g. https://example.com/foo.rpm.

Can we filter those out e.g. by rejecting slashes?  Explicit > implicit, and we can always loosen validation rules later.

> IOW, should butane ever try to recognize specific extensions, or should we just pass through to rpm-ostree install.

If certain extensions need additional configuration, we should add explicit sugar for those.  Otherwise, validation has no way to know that a particular extension will invoke additional logic in a future Butane release, so a newer Butane config could produce a subtly broken Ignition config with older Butane.

File: config/fcos/v1_5_exp/schema.go
Comment (bgilbert) [reply config/fcos/v1_5_exp/schema.go:43]: Followup in #316.

**Comment by cgwalters** [comment]
> Can we filter those out e.g. by rejecting slashes? Explicit > implicit, and we can always loosen validation rules later.

Actually I realized I was wrong, since the change to generate a treefile instead of forking `rpm-ostree install`, I think only package names/NEVRAs will be supported.  (The direct `https://example.com/foo.rpm` stuff lives in the client binary today)

**Comment by jmarrero** [comment]
https://github.com/coreos/butane/pull/304 & https://github.com/coreos/butane/pull/316 solve this issue.

**Pull request merged**
